Add correct string recommendations [I18N-ACTION-1216]

aphillips opened the pull request:
Added, edited, and rearranged requirements. Added a link to the Design Principles doc.

These edits depend on additions to the glossary in a separate pull request. More work is needed here, particularly discussing when the choose `USVString` (which we don't mention) and making clear how to work with byte strings/encodings.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/aphillips/bp-i18n-specdev/pull/82.html" title="Last updated on Dec 15, 2022, 5:16 PM UTC (ff699d7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/bp-i18n-specdev/82/4644204...aphillips:ff699d7.html" title="Last updated on Dec 15, 2022, 5:16 PM UTC (ff699d7)">Diff</a>

aphillips committed:
Add correct string recommendations [I18N-ACTION-1216]

Added, edited, and rearranged requirements. Added a link to the Design Principles doc.

These edits depend on additions to the glossary in a separate pull request. More work is needed here, particularly discussing when the choose `USVString` (which we don't mention) and making clear how to work with byte strings/encodings.

xfq commented:
The text looks good to me. I only have one suggestion: we should probably add a link to the specific section of the Design Principles doc, to make it easier to find.

And I agree that we should add an entry for "surrogate" to the glossary. I also mentioned it in https://github.com/w3c/i18n-glossary/issues/17

aphillips committed:
Further edits, added USVString recommendations

Added additional notes about the difference between `DOMString` and `USVString` and when to choose which. Added links to the new text in string-meta about byte strings. Added a note (per David's comment in telecon) linking to the encoding requirements.

aphillips commented on index.html:
```diff
@@ -1713,19 +1713,43 @@ <h3>Defining 'string'</h3>
+<p>The reason <code translate="no">DOMString</code> is preferred to <code translate="no">USVString</code> is that both [[DOM]] and the string types in JavaScript (and its derivatives, such as JSON) are defined in terms of UTF-16 code unit strings. Specifying <code translate="no">USVString</code> can result in inadvertently requiring an implementation to check for unpaired surrogates in cases wher ...
+
+<div class="req" id="char_string_usvstring">
+	<p class="advisement">When designing a web platform feature or API that operates on the internal values of strings, including indexing, iterating, transformation, or searching, the use of <a href="">USVString</a> is RECOMMENDED.</p>
```
add the link to usvstring (doh)

aphillips committed:
Address discussion in telecon of 2022-11-17

Added specific link to design-principles.

Copied in more text from D-P

Edited text throughout, including adding more color around 'byte string'

Added text about USVString and when to use it.

xfq commented on index.html:
```diff
@@ -1713,19 +1713,48 @@ <h3>Defining 'string'</h3>
-	<p class="advisement">The 'character string' definition SHOULD be used by most specifications.</p>
+<p>The type <code translate="no">USVString</code> defines strings as a sequence of Unicode <a>code points</a>. For strings whose most common algorithms operate on or process individual <a>code points</a>, or for operations which can’t handle surrogates in input, <code translate="no">USVString</code> should be used. For example, if your specification is defining a process that parses a string or t ...
+
+<p>In a <code translate="no">USVString</code>, isolated <a>surrogate</a> code points are invalid and implementations are required to replace any found in a string with the Unicode replacment character (<span class="codepoint" translate="no"><bdi lang="und">&#xFFFD;</bdi>[<span class="uname">U+FFFD REPLACEMENT CHARACTER</span>]</span>).</p>
```
```suggestion
<p>In a <code translate="no">USVString</code>, isolated <a>surrogate</a> code points are invalid and implementations are required to replace any found in a string with the Unicode replacment character (<span class="codepoint" translate="no"><bdi lang="und">&#xFFFD;</bdi> [<span class="uname">U+FFFD REPLACEMENT CHARACTER</span>]</span>).</p>
```

Add a space.

aphillips committed:
Update index.html per @xfq's suggestion

Co-authored-by: Fuqiao Xue <xfq@w3.org>

aphillips committed:
incorporate CLDR exemplar names

Addresses #84

This is a first cut, taking all of the current name list and transforming it to HTML. Some bugs remain and some names have no Latin transcription. Some names have no gender or region.

aphillips committed:
More fixing CLDR names

* Add gender data
* Add regions
* Add transcriptions to Latin

aphillips committed:
Add note on CLDR origins

aphillips committed:
fixing cldr data

* Split W3C and CLDR data apart
* Sort both tables
* Fix CLDR data
* Add text describing the sourcing of CLDR and better explaining the expected usage
* Moved additional information from the section bottom so that it isn't orphaned
* Converted notes to plain text inside sub-sub-sections for each table
* Fix table styles

aphillips committed:
Fix typo in local.css

Thanks to @xfq!

aphillips committed:
changes from 2022-12-15 telecon

* comment out CLDR names list

aphillips commented:
Merging per teleconference of 2022-12-15
